File: node/mods/rustscript/lib/opcodes/setarrayfield.js
/**
 * Purpose: SETARRAYFIELD opcode — write a field on each destination object
 * from a parallel source list (with last-value / scalar broadcasting).
 *
 * Language shape:
 *   script.destination  — context.* path to an array of objects
 *   script.source       — array, scalar, or tx collection ref
 *   script.field        — literal property name (not resolve_ref)
 *
 * destination[i][field] = source[min(i, source.len()-1)]
 *
 * Empty source arrays fail. Non-object destination elements fail.
 */

module.exports = {
  name: 'SETARRAYFIELD',
  description: `
Writes a named field on every object in a destination array.

source may be an array, a scalar (broadcast), or a transaction collection
(tx.from, tx.to, tx.path, tx.from.p2sh, tx.to.p2sh). When the source is
shorter than the destination, the last source value repeats.

field is a literal property name — it is not resolved as a reference.
`,
  exampleScript: {
    op: 'SETARRAYFIELD',
    destination: 'context.constitution',
    source: '__opcodes.importarray.successors',
    field: 'owner'
  },
  schema: {
    destination: 'string',
    source: 'reference',
    field: 'string'
  },

  execute(node, context) {
    if (!node || typeof node !== 'object' || !context || typeof context !== 'object') {
      return false;
    }

    const destination = node.destination;
    if (typeof destination !== 'string' || destination.length === 0) {
      return false;
    }

    if (!Object.prototype.hasOwnProperty.call(node, 'source')) {
      return false;
    }

    const field = node.field;
    if (typeof field !== 'string' || field.length === 0) {
      return false;
    }

    if (!destination.startsWith('context.')) {
      return false;
    }

    const path = destination.slice('context.'.length);
    if (!path || isForbiddenWritePath(path)) {
      return false;
    }

    const sourceValues = resolveSourceValues(node.source, context);
    if (!sourceValues || sourceValues.length === 0) {
      return false;
    }

    const dest = getContextPath(context, path);
    if (!Array.isArray(dest)) {
      return false;
    }
    if (!dest.every((e) => e && typeof e === 'object' && !Array.isArray(e))) {
      return false;
    }

    const lastIdx = sourceValues.length - 1;
    for (let i = 0; i < dest.length; i += 1) {
      const srcIdx = i < sourceValues.length ? i : lastIdx;
      dest[i][field] = deepClone(sourceValues[srcIdx]);
    }

    return true;
  }
};

function isForbiddenWritePath(path) {
  return (
    path === 'script' ||
    path.startsWith('script.') ||
    path === 'witness' ||
    path.startsWith('witness.') ||
    path === 'tx' ||
    path.startsWith('tx.') ||
    path === 'blk' ||
    path.startsWith('blk.')
  );
}

function pathSegments(path) {
  return String(path)
    .replace(/\[/g, '.')
    .replace(/\]/g, '')
    .split('.')
    .filter((s) => s.length > 0);
}

function getContextPath(root, path) {
  const segments = pathSegments(path);
  if (!segments.length) {
    return undefined;
  }
  let current = root;
  for (let i = 0; i < segments.length; i += 1) {
    const segment = segments[i];
    const asIndex = /^[0-9]+$/.test(segment) ? Number(segment) : null;
    if (asIndex !== null) {
      if (!Array.isArray(current) || asIndex >= current.length) {
        return undefined;
      }
      current = current[asIndex];
      continue;
    }
    if (
      !current ||
      typeof current !== 'object' ||
      Array.isArray(current) ||
      !Object.prototype.hasOwnProperty.call(current, segment)
    ) {
      return undefined;
    }
    current = current[segment];
  }
  return current;
}

function isP2shSlip(slip) {
  if (!slip || typeof slip !== 'object') {
    return false;
  }
  if (slip.slip_type === 'Bound' || slip.slip_type === 9) {
    return false;
  }
  const pk = slip.public_key || slip.publicKey;
  if (Array.isArray(pk) && pk[0] === 0) {
    return true;
  }
  if (pk && typeof pk === 'object' && pk[0] === 0) {
    return true;
  }
  return false;
}

function resolveSpecialCollection(name, context) {
  if (!context.tx || typeof context.tx !== 'object') {
    return null;
  }
  if (name === 'tx.from') {
    return Array.isArray(context.tx.from) ? context.tx.from : null;
  }
  if (name === 'tx.to') {
    return Array.isArray(context.tx.to) ? context.tx.to : null;
  }
  if (name === 'tx.path') {
    return Array.isArray(context.tx.path) ? context.tx.path : null;
  }
  if (name === 'tx.from.p2sh') {
    return Array.isArray(context.tx.from) ? context.tx.from.filter(isP2shSlip) : null;
  }
  if (name === 'tx.to.p2sh') {
    return Array.isArray(context.tx.to) ? context.tx.to.filter(isP2shSlip) : null;
  }
  return null;
}

function resolveSourceValues(source, context) {
  const resolved = resolveRef(context, source);
  if (Array.isArray(resolved)) {
    return resolved.length === 0 ? null : resolved;
  }

  if (typeof source === 'string') {
    switch (source) {
      case 'tx.from':
      case 'tx.to':
      case 'tx.path':
      case 'tx.from.p2sh':
      case 'tx.to.p2sh': {
        const values = resolveSpecialCollection(source, context);
        if (!values || values.length === 0) {
          return null;
        }
        return values;
      }
      default:
        break;
    }
  }

  return [resolved];
}

function deepClone(value) {
  return JSON.parse(JSON.stringify(value));
}

function resolveRef(root, ref) {
  if (typeof ref !== 'string') {
    return ref;
  }
  const parts = ref.split('.');
  let cursor = root;
  for (let i = 0; i < parts.length; i += 1) {
    const key = parts[i];
    if (
      !cursor ||
      typeof cursor !== 'object' ||
      !Object.prototype.hasOwnProperty.call(cursor, key)
    ) {
      return ref;
    }
    cursor = cursor[key];
  }
  return cursor;
}

File: rust/saito-core/src/core/consensus/scripting/opcodes/setarrayfield.rs
use crate::core::consensus::block::Block;
use crate::core::consensus::slip::{Slip, SlipType};
use crate::core::consensus::transaction::Transaction;
use serde_json::Value;

use super::super::script::resolve_ref;

/// SETARRAYFIELD — write a field on each destination object from a parallel
/// source list (with last-value / scalar broadcasting).
///
/// Script shape:
/// ```json
/// {
///   "op": "SETARRAYFIELD",
///   "destination": "context.constitution",
///   "source": "__opcodes.importarray.successors",
///   "field": "owner"
/// }
/// ```
///
/// For each destination index `i`:
/// `destination[i][field] = source[min(i, source.len()-1)]`
///
/// A scalar source broadcasts as a one-element list. Empty source arrays fail.
///
/// Destinations must be `context.*` paths that resolve to arrays of objects.
pub struct SetArrayField {
    pub name: String,
    pub description: String,
    pub script: String,
    pub schema: Value,
}

fn is_forbidden_write_path(path: &str) -> bool {
    path == "script"
        || path.starts_with("script.")
        || path == "witness"
        || path.starts_with("witness.")
        || path == "tx"
        || path.starts_with("tx.")
        || path == "blk"
        || path.starts_with("blk.")
}

/// Normalize `foo[0].bar` into dotted segments compatible with resolve_ref lookup.
fn path_segments(path: &str) -> Vec<String> {
    let normalized = path.replace('[', ".").replace(']', "");
    normalized
        .split('.')
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

fn get_context_path_mut<'a>(root: &'a mut Value, path: &str) -> Option<&'a mut Value> {
    let segments = path_segments(path);
    if segments.is_empty() {
        return None;
    }

    let mut current = root;
    for segment in &segments {
        if let Ok(index) = segment.parse::<usize>() {
            let arr = current.as_array_mut()?;
            current = arr.get_mut(index)?;
        } else {
            current = current.get_mut(segment)?;
        }
    }
    Some(current)
}

/// P2SH slip filter matching resolve_p2sh_slip_field / SETARRAY / ARRAYIFY.
fn is_p2sh_slip(slip: &Slip) -> bool {
    if slip.slip_type == SlipType::Bound {
        return false;
    }
    slip.public_key[0] == 0x00
}

fn slips_to_values(slips: &[Slip]) -> Option<Vec<Value>> {
    let value = serde_json::to_value(slips).ok()?;
    value.as_array().map(|a| a.clone())
}

fn p2sh_slips_to_values(slips: &[Slip]) -> Option<Vec<Value>> {
    let filtered: Vec<&Slip> = slips.iter().filter(|s| is_p2sh_slip(s)).collect();
    let value = serde_json::to_value(filtered).ok()?;
    value.as_array().map(|a| a.clone())
}

fn resolve_special_collection(name: &str, tx: Option<&Transaction>) -> Option<Vec<Value>> {
    match name {
        "tx.from" => tx.and_then(|t| slips_to_values(&t.from)),
        "tx.to" => tx.and_then(|t| slips_to_values(&t.to)),
        "tx.path" => tx.and_then(|t| {
            serde_json::to_value(&t.path)
                .ok()
                .and_then(|v| v.as_array().map(|a| a.clone()))
        }),
        "tx.from.p2sh" => tx.and_then(|t| p2sh_slips_to_values(&t.from)),
        "tx.to.p2sh" => tx.and_then(|t| p2sh_slips_to_values(&t.to)),
        _ => None,
    }
}

/// Resolve source to a non-empty list of values (array elements or one scalar).
fn resolve_source_values(
    source: &Value,
    context: &Value,
    tx: Option<&Transaction>,
    blk: Option<&Block>,
) -> Option<Vec<Value>> {
    let resolved = resolve_ref(source, context, tx, blk);
    if let Some(arr) = resolved.as_array() {
        if arr.is_empty() {
            return None;
        }
        return Some(arr.clone());
    }

    if let Some(s) = source.as_str() {
        match s {
            "tx.from" | "tx.to" | "tx.path" | "tx.from.p2sh" | "tx.to.p2sh" => {
                let values = resolve_special_collection(s, tx)?;
                if values.is_empty() {
                    return None;
                }
                return Some(values);
            }
            _ => {}
        }
    }

    // scalar broadcast
    Some(vec![resolved])
}

impl SetArrayField {
    pub fn validate(context: &mut Value, tx: Option<&Transaction>, blk: Option<&Block>) -> u8 {
        let destination = context["script"]["destination"]
            .as_str()
            .unwrap_or("")
            .to_string();
        if destination.is_empty() {
            return 0;
        }

        let Some(path) = destination.strip_prefix("context.") else {
            return 0;
        };
        let path = path.to_string();
        if path.is_empty() || is_forbidden_write_path(&path) {
            return 0;
        }

        let field = context["script"]["field"]
            .as_str()
            .unwrap_or("")
            .to_string();
        if field.is_empty() {
            return 0;
        }

        if context["script"].get("source").is_none() {
            return 0;
        }

        let Some(source_values) =
            resolve_source_values(&context["script"]["source"].clone(), context, tx, blk)
        else {
            return 0;
        };

        let Some(dest_value) = get_context_path_mut(context, &path) else {
            return 0;
        };

        let Some(dest_arr) = dest_value.as_array_mut() else {
            return 0;
        };

        if !dest_arr.iter().all(|e| e.is_object()) {
            return 0;
        }

        let last_idx = source_values.len() - 1;
        for (i, elem) in dest_arr.iter_mut().enumerate() {
            let src_idx = if i < source_values.len() { i } else { last_idx };
            let Value::Object(map) = elem else {
                return 0;
            };
            map.insert(field.clone(), source_values[src_idx].clone());
        }

        1
    }
}
